### PR TITLE
fix(coverage analysis): 'AppDomain' and named pipes are not available in all dotnet frameworks

### PR DESCRIPTION
**Pull request by dupdob**
Fix several incorrect log messages that where found while working on #781 . 
Note that my previous PR regarding Linq  mutators also appears as a difference, which is unexpected

### CODE REVIEW

**Comment by Mobrockers** [comment]
You made the changes on the linq mutator on the master branch, we then squashed the changes into one commit when accepting your merge request. Your master doesn't understand that the squash commit contains the same changes so it tries to reapply your previous commits. That's why they're showing up here.

**Comment by dupdob** [comment]
This PR is ready for review

File: src/Stryker.Core/Stryker.Core/Initialisation/InputFileResolver.cs [hunk 1]
```diff
@@ -177,7 +178,7 @@ private IEnumerable<string> ExtractProjectFolders(ProjectAnalyzerResult projectA
             var folders = new List<string>();
             var projectDirectory = _fileSystem.Path.GetDirectoryName(projectFilePath);
             folders.Add(projectDirectory);
-            if (!fullFramework)
+           // if (!fullFramework)
```
Comment (dupdob) [inline src/Stryker.Core/Stryker.Core/Initialisation/InputFileResolver.cs]: What is the point of this if? I disabled it while testing for compatibility with NFluent on Net 4.5 (otherwise, the project is seen as empty by Stryker).
Can you please give the reason why Stryker disregards share project for full framework? otherwise I suggest we get rid of this condition altogether.

File: src/Stryker.Core/Stryker.Core/Compiling/RollbackProcess.cs [hunk 2]
```diff
@@ -167,7 +167,7 @@ private SyntaxTree RemoveMutantIfStatements(SyntaxTree originalTree, ICollection
                 if (mutationIf == null)
                 {
                     var errorLocation = diagnostic.Location.GetMappedLineSpan();
-                    _logger.LogWarning("Stryker.NET encountered an compile error in {0} with message: {1} (Source code: {2})", errorLocation.Path, diagnostic.GetMessage(), brokenMutation);
+                    _logger.LogWarning($"Stryker.NET encountered an compile error in {errorLocation.Path} (at {errorLocation.StartLinePosition.Line}:{errorLocation.StartLinePosition.Character}) with message: {diagnostic.GetMessage()} (Source code: {brokenMutation})");
```
Comment (dupdob) [inline src/Stryker.Core/Stryker.Core/Compiling/RollbackProcess.cs]: I added more location information in case of build error so we can provide diagnostic faster.

File: src/Stryker.Core/Stryker.Core/Initialisation/InputFileResolver.cs (hunk 1 above)
Comment (Mobrockers) [reply src/Stryker.Core/Stryker.Core/Initialisation/InputFileResolver.cs]: At some point when shared projects were added we thought only dotnet core had shared projects. I guess we forgot to remove this check when we realized this isn't true. The if should be removed

**Comment by Mobrockers** [comment]
Hey @dupdob I created a fresh branch in which I cherry-picked your actual change commits so the history makes more sense for the PR, do you mind if I force push the changes to your branch? You can see the fixed history here: #810

**Comment by dupdob** [comment]
ok for the force push

File: src/Stryker.Core/Stryker.Core.UnitTest/Initialisation/InitialTestProcessTests.cs [hunk 3]
```diff
@@ -20,21 +21,33 @@ public void InitialTestProcess_ShouldThrowExceptionOnFail()
...
+            testRunnerMock.Setup(x => x.CaptureCoverage(false, false))
                 .Returns(new TestRunResult { Success = false });
-            testRunnerMock.Setup(x => x.DiscoverNumberOfTests()).Returns(0);
+            testRunnerMock.Setup(x => x.DiscoverNumberOfTests()).Returns(1);
 
             var exception = Assert.Throws<StrykerInputException>(() => _target.InitialTest(testRunnerMock.Object));
         }
 
+        [Fact]
+        public void InitialTestProcess_ShouldThrowWhenNoTest()
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core.UnitTest/Initialisation/InitialTestProcessTests.cs]: Why did you add this? I think it is fine if we keep running stryker on a project with 0 tests.

File: src/Stryker.Core/Stryker.Core/Initialisation/InitialTestProcess.cs [hunk 4]
```diff
@@ -31,6 +31,10 @@ public InitialTestProcess()
         public int InitialTest(ITestRunner testRunner)
         {
             var message = testRunner.DiscoverNumberOfTests() is var total && total == -1 ? "Unable to detect" : $"{total}";
+            if (total == 0)
```
Comment (Mobrockers) [inline src/Stryker.Core/Stryker.Core/Initialisation/InitialTestProcess.cs]: This is already gracefully handled by the mutationtestprocess here

 https://github.com/stryker-mutator/stryker-net/blob/cfc1e1cb3f0f9ef17bfab1d6a0a6e15cc7b01e16/src/Stryker.Core/Stryker.Core/MutationTest/MutationTestProcess.cs#L178

It will log a warning that none of the mutants were covered by a test and it will create the report with all created mutations. We do not need to throw on no tests found.

File: src/Stryker.Core/Stryker.Core/MutationTest/MutationTestProcess.cs [hunk 5]
```diff
@@ -63,12 +67,14 @@ public class MutationTestProcess : IMutationTestProcess
...
         public void Mutate()
         {
+            var preprocessorSymbols = _input.ProjectInfo.ProjectUnderTestAnalyzerResult.DefineConstants;
+
             _logger.LogDebug("Injecting helpers into assembly.");
             var mutatedSyntaxTrees = new List<SyntaxTree>();
-            var cSharpParseOptions = new CSharpParseOptions(_options.LanguageVersion);
-            foreach (var helper in CodeInjection.MutantHelpers)
+            var cSharpParseOptions = new CSharpParseOptions(_options.LanguageVersion, DocumentationMode.None, preprocessorSymbols: preprocessorSymbols);
+            foreach (var (name, code) in CodeInjection.MutantHelpers)
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core/MutationTest/MutationTestProcess.cs]: This is awesome! 🚀

**Comment by dupdob** [comment]
During my tests, ‘0 test found’ situation were linked to silent failure from VsTests. 
Maybe there is a better way to report this than my aggressive throw, but we should point this to users

Envoyé de mon iPhone

File: src/Stryker.Core/Stryker.Core/MutationTest/MutationTestProcess.cs [hunk 6]
```diff
@@ -79,12 +85,12 @@ public void Mutate()
...
                     options: cSharpParseOptions);
 
-                    _logger.LogDebug($"Mutating {file.Name}");
+                _logger.LogDebug($"Mutating {file.Name}");
                 // Mutate the syntax tree
                 var mutatedSyntaxTree = _orchestrator.Mutate(syntaxTree.GetRoot());
                 // Add the mutated syntax tree for compilation
                 mutatedSyntaxTrees.Add(mutatedSyntaxTree.SyntaxTree);
-
+                _logger.LogTrace($"Mutated {file.Name}:{Environment.NewLine}{mutatedSyntaxTree.ToFullString()}");
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core/MutationTest/MutationTestProcess.cs]: Do we need this extra logging? I'm afraid that logfiles from large projects will get even more unclear thanks to these enormous blocks of code. We already log quite some code during rollbacking. Do you think this will make things more clear?

File: src/Stryker.Core/Stryker.Core/MutationTest/MutationTestProcess.cs [hunk 7]
```diff
@@ -208,5 +214,18 @@ public StrykerRunResult Test(StrykerOptions options)
 
             return new StrykerRunResult(options, _input.ProjectInfo.ProjectContents.GetMutationScore());
         }
+
+        public TestCoverageInfos GetCoverage()
+        {
+            var (compat_noAppDomain, compat_noPipe) = _input.ProjectInfo.ProjectUnderTestAnalyzerResult.CompatibilityModes;
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core/MutationTest/MutationTestProcess.cs]: The variable names `compat_noAppDomain` and `compat_noPipe` are not very clear to me.

File: src/Stryker.Core/Stryker.Core.UnitTest/Initialisation/InputFileResolverTests.cs [hunk 8]
```diff
@@ -52,6 +53,38 @@ public InputFileResolverTests()
 </Project>";
         }
 
+        [Fact]
+        public void ProjectAnalyzerShouldDecodeFramework()
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core.UnitTest/Initialisation/InputFileResolverTests.cs]: I think these should be seperate test cases. Maybe xUnit's `[Theory]` could be used in this case.

File: src/Stryker.Core/Stryker.Core/Initialisation/ProjectInfo.cs [hunk 9]
```diff
@@ -50,27 +61,99 @@ public string AssemblyPath
...
+        {
+            get => _defineConstants ?? BuildDefineConstants();
+            set => _defineConstants = value;
+        }
+
+        private IList<string> BuildDefineConstants()
+        {
+            var constants = _analyzerResult?.GetProperty("DefineConstants")?.Split(";")?.ToList() ?? new List<string>();
+
+            var (compat_noAppDomain, compat_noPipe) = CompatibilityModes;
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core/Initialisation/ProjectInfo.cs]: These names don't make any sense to me. What do they contain?

File: src/Stryker.Core/Stryker.Core/Compiling/RollbackProcess.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dupdob) [reply src/Stryker.Core/Stryker.Core/Compiling/RollbackProcess.cs]: ok. It's just than when the number of parameters increases, interpolated strings keeps it readable.
Can revert

File: src/Stryker.Core/Stryker.Core/MutationTest/MutationTestProcess.cs (hunk 5 above)
Comment (dupdob) [reply src/Stryker.Core/Stryker.Core/MutationTest/MutationTestProcess.cs]: it's not me, just C# 7 :-)

File: src/Stryker.Core/Stryker.Core/MutationTest/MutationTestProcess.cs (hunk 6 above)
Comment (dupdob) [reply src/Stryker.Core/Stryker.Core/MutationTest/MutationTestProcess.cs]: I think it would be beneficial be able to access mutated code for mutation error analysis.
On the other hand, when testing, I realised that Trace entries are present when you request debug level. I think it would be better if we have a specific switch for that. Maybe correlating that with Devmode.
What to you think?

File: src/Stryker.Core/Stryker.Core/MutationTest/MutationTestProcess.cs (hunk 7 above)
Comment (dupdob) [reply src/Stryker.Core/Stryker.Core/MutationTest/MutationTestProcess.cs]: ok

File: src/Stryker.Core/Stryker.Core/Compiling/RollbackProcess.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (richardwerkman) [reply src/Stryker.Core/Stryker.Core/Compiling/RollbackProcess.cs]: We could also use `{errorLocation}` instead of `{0}` if that helps readability?

File: src/Stryker.Core/Stryker.Core/MutationTest/MutationTestProcess.cs (hunk 6 above)
Comment (richardwerkman) [reply src/Stryker.Core/Stryker.Core/MutationTest/MutationTestProcess.cs]: During development you could also get the mutated code using debugging right? A devmode switch could be better here. If you think it is beneficial we can keep it with a switch :+1:

Comment (Mobrockers) [reply src/Stryker.Core/Stryker.Core/MutationTest/MutationTestProcess.cs]: When logging to file, all logging is written to the file. This is intended behaviour. If you mean that you see trace logging in the console when requesting debug logging then that is a bug and that should not be happening.

Comment (dupdob) [reply src/Stryker.Core/Stryker.Core/MutationTest/MutationTestProcess.cs]: Ok, added devmode has a supplemental condition
